docs/source/testing.rst

```diff
@@ -0,0 +1,218 @@
+application. Typically python execution stops at this point and control is
+passed to the underlying event loop. In a test context though, we need
+direct access to that event loop. To simulate and test GUI interactions we need
+to be able to post events to the event loop, process them manually, and then
+make assertions about the desired behavior. Even further, the event loop is
+global state. Therefor, great care needs to be taken in each test to pick up
+after itself to avoid interactions with other tests. This is still necessary
+even if the test fails. 
+
+Pyface provides a few utilities that are useful in this process.  Namely,
+:class:`~pyface.util.GuiTestAssistant` and
+:class:`~pyface.util.ModalDialogTester`.
+
+
+GuiTestAssistant
+================
+
+.. note::
+
+   GuiTestAssistant is currently only available on Qt.
+
+:class:`GuiTestAssistant` is a mixin class intended to to augment
+:class:`python:unittest.TestCase`. In fact, it has
+:class:`~traits.testing.unittest_tools.UnittestTools` from traits as a base
+class and thus gives access to methods like
+:meth:`~traits.testing.unittest_tools.UnittestTools.assertTraitChanges`. See the
+`Traits Testing documentation <https://docs.enthought.com/traits/traits_user_manual/testing.html#testing>`_
+for more. 
+
+:class:`GuiTestAssistant` holds a reference to a :class:`pyface.gui.GUI` object
+(for api details see the interface :class:`~pyface.i_gui.IGUI`) which is what
+gives the low level access to the event loop. :class:`pyface.gui.GUI` provides methods such as
+:meth:`~pyface.i_gui.IGUI.start_event_loop`,
+:meth:`~pyface.i_gui.IGUI.stop_event_loop`,
+:meth:`~pyface.i_gui.IGUI.process_events`,
+:meth:`~pyface.i_gui.IGUI.invoke_later`, and
+:meth:`~pyface.i_gui.IGUI.set_trait_later`. This is accessible via the
+:attr:`gui` attribute on :class:`GuiTestAssistant`.
+
+What :class:`GuiTestAssistant` provides that is novel, is effectively more
+security to make sure your tests clean up after themselves.  For example,
+:class:`GuiTestAssistant` provides standard :meth:`setUp` and :meth:`tearDown`
+methods which try to clean up existing UI state and empty the event loop even
+if a test fails.  In addition, the methods typically have timouts so that the
+test will fail rather than blocking forever in the case something has gone
+wrong. Effectively, the class aims to remember to do the overhead to ensure
+your tests don't cause trouble, and at the same time give you the low level
+event loop access needed to write your GUI tests.
+
+
+This class provides the following methods (some of them being context managers):
+
+- :meth:`event_loop`
+  
+    Takes an integer ``repeat`` parameter and artificially replicates the event
+    loop by calling :meth:`sendPostedEvents` and :meth:`processEvents` ``repeat``
+    number of times.
+
+- :meth:`event_loop_until_condition`
+
+    Runs the real Qt event loop until the provided condition evaluates to True.
```

Comment:
One current usage difficulty with `GuiTestAssistant` is figuring out which methods are context managers: some of the methods are designed to be used a context managers, and if used without a `with` statement do nothing instead. (We've had more than one bug in real code as a result of this.)

It would be worth clearly noting for each method whether its return value is intended to be used as a context manager or not.